### PR TITLE
2017-07-10 - Robert Tisma - Detail Example of Writing CSV Files

### PR DESCRIPTION
**Pull request by rtisma**

### CODE REVIEW

File: _posts/robert_tisma/2017-07-10-writing-csv-files.md [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+
+A few months ago, I worked on a project that involved writing model objects to CSV files in a particular order with specific header names. Since classes defining the models were located in an external library, the field names could not be renamed to match the required headers. I then set out on a mission to analyze how I could approach this problem using different libraries, such as SuperCSV, OpenCSV, Jackson and even my very own simple implementation. In this blog post, I will be defining the problem, describing the design used to solve the problem, and analyzing the results of different implementations. 
+
+## Objective
+
+The objective is to write a collection of [beans](https://en.wikipedia.org/wiki/Plain_old_Java_object#JavaBeans) 
+or [POJOs](https://en.wikipedia.org/wiki/Plain_old_Java_object) (Plain Old Java Object - which is a mutable Java class that contains an empty, argument-less constructor, with a getter and setter method for each field) to CSV files using several external libraries, while adhering to the following constraints:
+
+- __Rule 1:__ _Ordered Headers_ - the number of headers and their order may very from how they are defined in the model.
```
Comment (Fgerthoffert) [inline _posts/robert_tisma/2017-07-10-writing-csv-files.md]: may **vary** from

File: _posts/robert_tisma/2017-07-10-writing-csv-files.md [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+1. __Explicit Schema Strategy__: converting a bean to a `String[]` by hardcoding method calls
+2. __Lambda Schema Strategy__: converting a bean to a `String[]` by calling Java 8 method references that point to the getter methods of the bean.
+
+Ultimately, the software must be maintainable and human readable. The goal is to write loosely coupled and highly cohesive software, that not only adheres to the [SOLID](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)) design principles but is also self-documenting.
+
+## Background
+
+### What is the CSV format and RFC-4180?
+
+The term CSV stands for "Comma Separated Value", and represents a loosely defined format for representing tabular data using plain text, where commas separate field values (columns) and newlines separate records (rows). In fact, CSV files are not limited to the comma character being the delimiter, for instance, a TSV (Tab Separated Value) file is a CSV file that uses the tab character instead of the comma character. CSV is commonly used as a data exchange format for scientific and business applications and is popular for its simplicity and for being human readable.
```
Comment (Fgerthoffert) [inline _posts/robert_tisma/2017-07-10-writing-csv-files.md]: Just a personal opinion but I would rephrase the second sentence with: 
*Plain text representation of tabular data is not limited to CSV format, TSV (Tab Separated Value) provides an alternative, using the tab character instead of the comma character.*

File: _posts/robert_tisma/2017-07-10-writing-csv-files.md [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+
+In contrast to the previous test, this test uses the tab character `\t` as the separator, __with quotes__ and uses the regular order defined in `PersonSchema`. The JUnit test asserts that all configurations match the expected output CSV file, except for the `JACKSON_BEAN` configuration. By observing the `JACKSON_BEAN.output.csv` file contents:
+
+``` 
+"personId"      "personFirstName"       "personLastName"        "personAge"
+1               "Denis"                 "Ritchie"               70
+2               "Frank"                 "Rosenblatt"            43
+3               "Alex"                  "Murphy"                37
+```
+it is clear that the `jackson-dataformat-csv` library adds quotes to string fields, and does not add quotes to numeric values. If this external library is used to read AND write CSV files, then it is perfectly fine to use. However, if it is used to write CSV files, and then another library reads that CSV file, a error may occur when the library parses the unquoted numeric field.
```
Comment (Fgerthoffert) [inline _posts/robert_tisma/2017-07-10-writing-csv-files.md]: a**n** error may occur

**Review by Fgerthoffert — COMMENTED** [review]
Great Article !

**Comment by cheapsteak** [comment]
Post seems to be missing front-matter
Please reference other posts, e.g. https://github.com/OICR/softeng-blog/blame/gh-pages/_posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md#L1-L20

**Comment by cheapsteak** [comment]
Without front-matter, it ends up looking like this - 
image

image

Would recommend building locally to preview how it looks. Feel free to reach out to this site's architect @SidSyrus for help if you run into issues setting up a local instance

**Comment by rtisma** [comment]
@Fgerthoffert  i fixed the issues with sids help, and updated the formatting.
